controllers/C_Document.class.php

```diff
@@ -801,18 +796,20 @@ public function retrieve_action(string $patient_id = null, $document_id, $as_fil
         if (file_exists($temp_url)) {
             $url = $temp_url;
         }
-        //fire a remote call to see if the file is stored somewhere else
-        $s3Key = explode("//", $temp_url); //split the url to get the s3 key
-        // if the s3 key isn't empty then ...
-        if (!empty($s3Key[1])) {
-            $retrieveOffsiteDocument = new PatientRetrieveOffsiteDocument("/" . $s3Key[1]);
-            $this->eventDispatcher->dispatch($retrieveOffsiteDocument, PatientRetrieveOffsiteDocument::REMOTE_DOCUMENT_LOCATION);
-            //this is for the s3 bucket module. If the file is not found locally, it will be found remotely
-            if ($retrieveOffsiteDocument->getOffsiteUrl() != null) {
-                header('Content-Description: File Transfer');
-                header("Location: " . $retrieveOffsiteDocument->getOffsiteUrl());
-                exit;
-            }
+
+        $retrieveOffsiteDocument = new PatientRetrieveOffsiteDocument($temp_url);
+        $updatedOffsiteDocumentEvent = $GLOBALS['kernel']->getEventDispatcher()->dispatch(
+            $retrieveOffsiteDocument,
+            PatientRetrieveOffsiteDocument::REMOTE_DOCUMENT_LOCATION
+        );
+        // if a module writer has an independent offsite storage mechanism used this accomdoates that.
+        // If the file is not found locally, it will be found remotely.  Systems like s3, blob stores, etc, can
+        // be tied and and use those urls.  Note NO security is handled here so any kind of security mechanism must
+        // be handled on the receiving end's URL (s3/azure for example use signed urls with signature verification)
+        if (!empty($updatedOffsiteDocumentEvent) && $updatedOffsiteDocumentEvent->getOffsiteUrl() != null) {
```

Comment:
@adunsulag can you clarify what this part is doing?

        if (!empty($updatedOffsiteDocumentEvent)

Reply:
The first clause protects OpenEMR if someone in their module decides to return null for their event in the dispatched event handler. I think Symfony may catch this and just return the passed in argument but I get a little paranoid at times when interfacing with third party software.  I probably should have done a type check using instanceof instead of !empty which I can go ahead and do that.